### PR TITLE
add automatic type conversion support to ConvertProcessor

### LINKED ISSUE #17139: Add auto detection to convert processor
The `ingest` processor is able to convert string values to either a boolean, integer or float value.
This processor needs to be configured in a strict manner. For example `field_a` need to be converted to an integer. If the field doesn't exist or can't be its value can't converted to an integer this processor fails.

The idea is here the add an `auto` convert type option that tries to convert strings to either a integer/float or a boolean value based on simple rules:
* true | false -> boolean 
* [-+]?[0-9] -> integer
* [-+]?[0-9]*\.?[0-9]* -> float

If the `auto` convert type is unable to convert than it just does nothing and leaves that field untouched.

In order to auto detect a number fields it would be helpful if the `field` option would support wildcards too. 
For all fields that match the convert processor will convert the values based on the configured convert type. If no fields match with the wildcard pattern than the `convert` processor does nothing.

I think if this gets implemented then this can supersede number detections in mappings?

### PR DESCRIPTION
**Pull request by talevy**
closes: https://github.com/elastic/elasticsearch/issues/17139

todos:
- [x] finalize `target_field` behavior
- [x] update docs

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/ingest/processor/ConvertProcessor.java [hunk 1]
```diff
@@ -73,6 +73,23 @@ public Object convert(Object value) {
             public Object convert(Object value) {
                 return value.toString();
             }
+        }, AUTO {
+            @Override
+            public Object convert(Object value) {
+                if (!(value instanceof String)) {
+                   return value;
+                }
+                try {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/processor/ConvertProcessor.java]: I like the fact it is reusing the current convert types!

File: core/src/test/java/org/elasticsearch/ingest/processor/ConvertProcessorTests.java [hunk 2]
```diff
@@ -265,4 +266,85 @@ public void testConvertNullField() throws Exception {
             assertThat(e.getMessage(), equalTo("Field [field] is null, cannot be converted to type [" + type + "]"));
         }
     }
+
+    private Object randomValueFromTypes(Type... types) {
```
Comment (talevy) [inline core/src/test/java/org/elasticsearch/ingest/processor/ConvertProcessorTests.java]: unused now. forgot to remove. will get on that

**Comment by talevy** [comment]
@martijnvg updated with `target_field` and docs

File: core/src/test/java/org/elasticsearch/ingest/processor/ConvertProcessorTests.java (hunk 2 above)
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/ingest/processor/ConvertProcessorTests.java]: It is still unused, so it can be removed?

**Comment by martijnvg** [comment]
left one minor comment, LGTM otherwise.

File: core/src/test/java/org/elasticsearch/ingest/processor/ConvertProcessorTests.java (hunk 2 above)
Comment (talevy) [inline core/src/test/java/org/elasticsearch/ingest/processor/ConvertProcessorTests.java]: ah silly me, yes. will remove

**Pull request merged**